single-node-refactor/src/Solvers/SGH_solver/include/contact.h

```diff
@@ -0,0 +1,148 @@
+     * @param vz_max: ||        ||        ||        ||        ||        ||
+     * @param ax_max: absolute maximum x acceleration across all nodes in the patch
+     * @param ay_max: ||        ||        ||        ||        ||        ||
+     * @param az_max: ||        ||        ||        ||        ||        ||
+     * @param dt: time step
+     * @param bounds: array of size 6 that will contain the maximum and minimum components of the capture box in the
+     *                following order (xc_max, yc_max, zc_max, xc_min, yc_min, zc_min)
+     */
+    void capture_box(const double &vx_max, const double &vy_max, const double &vz_max,
+                     const double &ax_max, const double &ay_max, const double &az_max,
+                     const double &dt, CArrayKokkos<double> &bounds) const;
+
+};
+
+struct contact_patches_t
+{
+    CArrayKokkos<contact_patch_t> contact_patches;  // patches that will be checked for contact
+    CArrayKokkos<size_t> patches_gid;  // global patch ids
+    CArrayKokkos<size_t> nodes_gid;  // global node ids
+    size_t num_contact_patches;  // total number of patches that will be checked for contact
+
+    /*
+     * Sets up the contact_patches array
+     *
+     * @param mesh: mesh object
+     * @param bdy_contact_patches: global ids of patches that will be checked for contact
+     * @param nodes: node object that contains coordinates and velocities of all nodes
+     */
+    void initialize(const mesh_t &mesh, const CArrayKokkos<size_t> &bdy_contact_patches, const node_t &nodes);
+
+    /*
+     * Here is a description of each array below:
+     *
+     * nbox: An array consisting of the number of nodes that a bucket contains. For example, nbox[0] = 2 would mean
+     *       that bucket 0 has 2 nodes.
+     * lbox: An array consisting of the bucket id for each node. For example, lbox[5] = 12 would mean that node 5 is in
+     *       bucket 12.
+     * nsort: An array consisting of sorted nodes based off the bucket id.
+     * npoint: An array consisting of indices into nsort where each index is the starting location. For example,
+     *         nsort[npoint[5]] would return the starting node in bucket 5. This is a means for finding the nodes given
+     *         a bucket id.
+     *
+     * With the above data structure, you could easily get the nodes in a bucket by the following pythonic syntax:
+     * nsort[npoint[bucket_id]:npoint[bucket_id] + nbox[bucket_id]]
+     *
+     * Buckets are ordered by propagating first in the x direction, then in the y direction, and finally in the z.
+     */
+    CArrayKokkos<size_t> nbox;  // Size nb buckets
+    CArrayKokkos<size_t> lbox;  // Size n nodes (n is the total number of nodes being checked for penetration)
+    CArrayKokkos<size_t> nsort;  // Size n nodes
+    CArrayKokkos<size_t> npoint;  // Size nb buckets
+
+    static double bs;  // bucket size (defined as 1.001*min_node_distance) todo: consider changing it back to 0.999
```

Comment:
What are the implications of 1.001 vs 0.999?

Reply:
If the bucket size exactly matches the element size, then it will capture more nodes than needed because many will land on an edge. Making the size slightly different will fix this issue. I left that comment with 0.999 because I think it's better to have the size a little bit smaller than the element size. This is issue is only prevalent in structured meshes and the first iteration. After that first iteration, the elements have deformed, and it's unlikely that nodes will land exactly on the edge of a bucket.